packages/ra-keycloak/src/authProvider.ts

```diff
@@ -0,0 +1,118 @@
+import { AuthProvider } from 'react-admin';
+import Keycloak, { KeycloakTokenParsed } from 'keycloak-js';
+import jwt_decode from 'jwt-decode';
+
+export type PermissionsFunction = (decoded: KeycloakTokenParsed) => boolean;
+
+/**
+ * An authProvider which handles authentication via the Keycloak server.
+ *
+ * @example
+ * import dataProvider from "./dataProvider";
+ * import i18nProvider from "./i18nProvider";
+ * import Layout from "./Layout";
+ * import posts from "./posts";
+ *
+ * import Keycloak, { KeycloakConfig } from "keycloak-js";
+ * import { keycloakAuthProvider } from "ra-keycloak/authProvider";
+ *
+ * const isPermitted = (decoded: KeycloakTokenParsed): boolean => {
+ *   if (!decoded.resource_access) {
+ *     return false;
+ *   }
+ *   const admin = decoded.resource_access["$KEYCLOAK_CLIENT_ID"].roles.find(
+ *     (role) => role === "admin"
+ *   );
+ *   return !!admin;
+ * };
+ *
+ * const App = () => {
+ *   const [authProvider, setAuthProvider] = React.useState<AuthProvider | null>(
+ *     null
+ *   );
+ *
+ *   React.useEffect(() => {
+ *     async function startAuthProvider() {
+ *       const config: KeycloakConfig = {
+ *          url: "$KEYCLOAK_URL",
+ *          realm: "$KEYCLOAK_REALM",
+ *          clientId: "$KEYCLOAK_CLIENT_ID",
+ *       };
+ *
+ *       const keycloak = new Keycloak(config);
+ *       await keycloak.init({ onLoad: "login-required" });
+ *       const authProvider = keycloakAuthProvider(keycloak, {onPermissions: isPermitted});
+ *       setAuthProvider(authProvider);
+ *     }
+ *     if (authProvider === null) {
+ *       startAuthProvider();
+ *     }
+ *   }, [authProvider]);
+ *
+ *   // hide the admin until the data provider is ready
+ *   if (!authProvider) return <p>Loading...</p>;
+ *
+ *   return (
+ *     <Admin
+ *       authProvider={authProvider}
+ *       dataProvider={dataProvider}
+ *       i18nProvider={i18nProvider}
+ *       title="Example Admin"
+ *       layout={Layout}
+ *     >
+ *       <>
+ *         <Resource name="posts" {...posts} />
+ *         {....}
+ *       </>
+ *     </Admin>
+ *   );
+ * };
+ *
+ * @param {PermissionsFunction} onPermissions the function to decide if the authenticated user has the right to access to a specific resource.
+ * @returns {object} the authProvider object used by React-Admin.
+ */
+export const keycloakAuthProvider = (
+    client: Keycloak,
+    options: {
+        onPermissions?: PermissionsFunction;
+        redirectLogin?: string;
+        redirectLogout?: string;
+    } = {}
+): AuthProvider => ({
+    async login() {
+        return client.login({
+            redirectUri:
+                options.redirectLogin ?? window.location.origin + '/#/',
```

Comment:
`window.location.origin + '/#/'` -> this is only true if you use a HashRouter. We should find a more generic way to deal with redirection